lib/url.js

```diff
@@ -75,355 +88,192 @@ const slashedProtocol = {
   'gopher:': true,
   'file:': true
 };
+
+const _autoEscapeCharacters = makeAsciiTable(_autoEscape.map(function(v) {
+  return v.charCodeAt(0);
+}));
+
+// Characters that are never ever allowed in a hostname.
+// Note that any invalid chars are also handled, but these
+// are the ones that are *expected* to be seen, so we fast-path them.
+const _hostEndingCharacters = makeAsciiTable([
+  '#',
+  '?',
+  '/',
+  '\\'
+].map(function(v) {
+  return v.charCodeAt(0);
+}));
+// If these characters end a host name, the path will not be prepended a /.
+const _hostEndingCharactersNoPrependSlash = makeAsciiTable([
+  '<',
+  '>',
+  '"',
+  '`',
+  ' ',
+  '\r',
+  '\n',
+  '\t',
+  '{',
+  '}',
+  '|',
+  '^',
+  '`',
+  '\'',
+  '%',
+  ';'
+].map(function(v) {
+  return v.charCodeAt(0);
+}));
+
 const querystring = require('querystring');
 
+const noSerializePattern = /^(?:_protocol|_port|_href|_query|_prependSlash)$/;
+
 function urlParse(url, parseQueryString, slashesDenoteHost) {
   if (url instanceof Url) return url;
 
-  var u = new Url;
+  var u = new Url();
   u.parse(url, parseQueryString, slashesDenoteHost);
   return u;
 }
 
-Url.prototype.parse = function(url, parseQueryString, slashesDenoteHost) {
-  if (typeof url !== 'string') {
-    throw new TypeError("Parameter 'url' must be a string, not " + typeof url);
```

Comment:
Care to switch the quotes around? While not official yet (#1220), I think we'll agree on double-quoting parameters in errors inside single-quoted error strings.